File: code/chap07/datasource_mongodb_reader.py
from __future__ import print_function 
import sys 
from pyspark.sql import SparkSession 
from pyspark.sql import Row

#-----------------------------------------------------
# Read MongoDB Collection and create a DataFrame
# Input: MongoDB Collection URI
#------------------------------------------------------
# Input Parameters:
#    MongoDB Collection URI
#-------------------------------------------------------
# @author Mahmoud Parsian
#-------------------------------------------------------
def main():

    #if len(sys.argv) != 2:  
    #    print("Usage: datasource_mongodb_reader.py <collection>", file=sys.stderr)
    #    exit(-1)

    # read name of mongodb collection URI
    mongodb_collection_uri = sys.argv[1]
    print("mongodb_collection_uri : ", mongodb_collection_uri)

    # create an instance of SparkSession
    spark = SparkSession\
        .builder\
        .appName("datasource_mongodb_reader")\
        .config("spark.mongodb.input.uri", mongodb_collection_uri)\
        .config("spark.mongodb.output.uri", mongodb_collection_uri)\
        .getOrCreate()
    #
    print("spark=",  spark)

    # Assign the collection to a DataFrame with spark.read() 
    # from within the pyspark shell.
    df = spark.read.format("com.mongodb.spark.sql.DefaultSource").load()
    #
    print("df.count() = " , df.count())
    #
    print("df.collect() = " , df.collect())
    #
    df.show(10, truncate=False)        
    #
    # Spark samples the records to infer the schema of the collection.
    df.printSchema()
 
    # done!
    spark.stop()
#end-def

if __name__ == '__main__':
    main()

File: code/chap07/scala/src/main/scala/org/data/algorithms/spark/ch07/DatasourceMongodbReader.scala
package org.data.algorithms.spark.ch07

import org.apache.spark.sql.SparkSession
import org.apache.spark.sql.functions.col
/**
 *-----------------------------------------------------
 * Read Content of MongoDB to Dataframe
 * Input: MongoDB Collection uri (mongodb://127.0.0.1/test.myCollection)
 *------------------------------------------------------
 * Input Parameters:
 *    MongoDB Collection uri (mongodb://127.0.0.1/test.myCollection)
 *-------------------------------------------------------
 *
 * @author Deepak Kumar
 *-------------------------------------------------------
 */
object DatasourceMongodbReader {
  def main(args: Array[String]): Unit = {

    //read name of mongodb collection URI
    val mongoDBCollectionURI = args(0)
    println(s"MongoDBCollectionURI = ${mongoDBCollectionURI}")
    //create an instance of SparkSession
    val spark =
      SparkSession.
        builder().
        appName("DatasourceMongodbReader").
        master("local[*]").
        config("spark.mongodb.input.uri",mongoDBCollectionURI).
        config("spark.mongodb.output.uri",mongoDBCollectionURI).
        getOrCreate()

    //Assign the collection to a DataFrame with spark.read()
    val columnNames = List("name", "city", "age")
    val df = spark.read.format("com.mongodb.spark.sql.DefaultSource").load.select(columnNames.map(m=>col(m)):_*)
    //To read the contents of the DataFrame, use the show() method.
    df.show(10,truncate = false)
    println(s"df.count(): ${df.count()}")
    println(s"df.collect(): ${df.collect().mkString("",",","")}")
    //Spark samples the records to infer the schema of the collection.
    df.printSchema()

    //Done.
    spark.stop()

  }

}

/*
df.show
+-------+---------+---+
|name   |city     |age|
+-------+---------+---+
|Betty  |Ames     |78 |
|Gandalf|Cupertino|60 |
|Brian  |Stanford |77 |
|Alex   |Ames     |50 |
|Thorin |Sunnyvale|95 |
+-------+---------+---+
 */